dags/tmf/bikeshare/trips.py

```diff
@@ -0,0 +1,197 @@
+    TABLEAU_SITENAME = "echo"
+    TABLEAU_VERSION = "2.7"
+    TABLEAU_USER = Variable.get("BIKESHARE_TABLEAU_USER")
+    TABLEAU_PASSWORD = Variable.get("BIKESHARE_TABLEAU_PASSWORD")
+    TRIP_TABLE_VIEW_ID = "7530c937-887e-42da-aa50-2a11d279bf51"
+    logging.info("Authenticating with Tableau")
+    tableau_auth = tableauserverclient.TableauAuth(
+        TABLEAU_USER, TABLEAU_PASSWORD, TABLEAU_SITENAME,
+    )
+    tableau_server = tableauserverclient.Server(TABLEAU_SERVER, TABLEAU_VERSION)
+    tableau_server.auth.sign_in(tableau_auth)
+
+    # Get the Trips table view. This is a view specifically created for
+    # this DAG. Tableau server doesn't allow the download of underlying
+    # workbook data via the API (though one can from the UI). This view
+    # allows us to get around that.
+    logging.info("Loading Trips view")
+    all_views, _ = tableau_server.views.get()
+    view = next(v for v in all_views if v.id == TRIP_TABLE_VIEW_ID)
+    if not view:
+        raise Exception("Cannot find the trips table!")
+    tableau_server.views.populate_csv(view)
+    df = pandas.read_csv(
+        io.BytesIO(b"".join(view.csv)),
+        parse_dates=["Start Datetime", "End Datetime"],
+        thousands=",",
+        dtype={"visible_id": str, "end_station": str, "start_station": str},
+    )
+
+    # The data has a weird structure where trip rows are duplicated, with variations
+    # on a "Measure" column, containing trip length, duration, etc. We pivot on that
+    # column to create a normalized table containing one row per trip.
+    logging.info("Cleaning Data")
+    df = pandas.merge(
+        df.set_index("Trip ID")
+        .groupby(level=0)
+        .first()
+        .drop(columns=["Measure Names", "Measure Values"]),
+        df.pivot(index="Trip ID", columns="Measure Names", values="Measure Values"),
+        left_index=True,
+        right_index=True,
+    ).reset_index()
+    df = df.rename(
+        {
+            n: n.lower().strip().replace(" ", "_").replace("(", "").replace(")", "")
+            for n in df.columns
+        },
+        axis="columns",
+    )
+    check_columns(bike_trips, df)
+
+    # Upload the final dataframe to Postgres. Since pandas timestamps conform to the
+    # datetime interfaace, psycopg can correctly handle the timestamps upon insert.
+    logging.info("Uploading to PG")
+    engine = PostgresHook.get_hook(POSTGRES_ID).get_sqlalchemy_engine()
+    insert = sqlalchemy.dialects.postgresql.insert(bike_trips).on_conflict_do_nothing()
+    conn = engine.connect()
+    conn.execute(insert, *df.to_dict(orient="record"))
+
+
+def load_s3_data(**kwargs):
+    """
+    Load the table from PG and upload it as a parquet to S3.
+    """
+    bucket = kwargs.get("bucket")
+    name = kwargs.get("name", "bikeshare_trips.parquet")
+    if bucket:
+        logging.info("Uploading data to s3")
+        engine = PostgresHook.get_hook(POSTGRES_ID).get_sqlalchemy_engine()
```

Comment:
this is a good method, we should stub it out for my janky method in a few existing dags including `dockless-elt` and `311-to-care`